### PR TITLE
 Rule: Array.reverse should not be used misleadingly

### LINKED ISSUE #612: Rule: "Array.reverse" should not be used misleadingly
Implements [RSPEC-4043](https://jira.sonarsource.com/browse/RSPEC-4043)

### PR DESCRIPTION
**Pull request by inverno**
Fixes #612

### CODE REVIEW

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    List<IdentifierTree> bindingIdentifiers = bindingElement.left().bindingIdentifiers();
+    return bindingIdentifiers.get(bindingIdentifiers.size() - 1);
+  }
+
+  private boolean sameSymbol(Symbol objectSymbol, IdentifierTree lastIdentifier) {
+    Optional<Symbol> bindingSymbol = lastIdentifier.symbol();
+    return bindingSymbol.map(symbol -> symbol.equals(objectSymbol)).orElse(false);
+  }
+
+  private boolean isArray(Symbol symbol, ProgramState currentState) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: MAJOR Make "isArray" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    return currentState.getConstraint(symbol).isStricterOrEqualTo(Constraint.ARRAY);
+  }
+
+  private boolean isReverseMethod(ExpressionTree property) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: MAJOR Make "isReverseMethod" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    if (element.parent().is(Tree.Kind.ARGUMENT_LIST)) {
+      return true;
+    }
+    if (element.parent().parent() != null && element.parent().parent().is(Tree.Kind.ARROW_FUNCTION)) {
+      return true;
+    }
+    return false;
+  }
+
+  private IdentifierTree retrieveAssignmentIdentifier(Tree element) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: MAJOR Make "retrieveAssignmentIdentifier" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        MemberExpressionTree memberExpression = (MemberExpressionTree) callExpression.callee();
+        ExpressionTree object = memberExpression.object();
+        if (!isReverseMethod(memberExpression.property())) {
+          return;
+        }
+        if (!object.is(Tree.Kind.IDENTIFIER_REFERENCE)) {
+          return;
+        }
+        ((IdentifierTree) object).symbol().ifPresent(objectSymbol -> {
+          if (isArray(objectSymbol, currentState) && isBeingPassedElsewhere(objectSymbol, element))
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: MINOR Missing curly brace. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00121)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    return false;
+  }
+
+  private IdentifierTree retrieveAssignmentIdentifier(Tree element) {
+    InitializedBindingElementTree bindingElement = (InitializedBindingElementTree) element.parent();
+    List<IdentifierTree> bindingIdentifiers = bindingElement.left().bindingIdentifiers();
+    return bindingIdentifiers.get(bindingIdentifiers.size() - 1);
+  }
+
+  private boolean sameSymbol(Symbol objectSymbol, IdentifierTree lastIdentifier) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: MAJOR Make "sameSymbol" a "static" method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2325)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+      AssignmentExpressionTree assignment = (AssignmentExpressionTree) element.parent();
+      if (assignment.variable().is(Tree.Kind.IDENTIFIER_REFERENCE)) {
+        IdentifierTree identifier = (IdentifierTree) assignment.variable();
+        return !sameSymbol(objectSymbol, identifier);
+      }
+    }
+    if (element.parent().is(Tree.Kind.ARGUMENT_LIST)) {
+      return true;
+    }
+    if (element.parent().parent() != null && element.parent().parent().is(Tree.Kind.ARROW_FUNCTION)) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: MINOR Replace this if-then-else statement by a single return statement. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1126)

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* ![MAJOR][MAJOR] 4 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java (hunk 4 above)
Comment (vilchik-elena) [reply javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: @inverno any plans to add braces here?:)

File: javascript-checks/src/test/resources/checks/ArrayReverse.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  const b = a.reverse(); // Noncompliant {{Move this array "reverse" operation to a separate statement.}}
+//          ^^^^^^^^^^^
+
+  a.slice().reverse().forEach(() => {}); // OK, there is `slice`
+
+  a.reverse(); // OK
+
+  a.map(() => true).reverse(); // OK
+
+  a = a.reverse(); // OK, just a minor code-smell
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/ArrayReverse.js]: @inverno it's not yet a code smell. ticket?

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            this.addUniqueIssue(element, MESSAGE);
+        });
+
+      }
+    }
+  }
+
+  private boolean isBeingPassedElsewhere(Symbol objectSymbol, Tree element) {
+    if (element.parent().is(Tree.Kind.INITIALIZED_BINDING_ELEMENT)) {
+      return !sameSymbol(objectSymbol, retrieveAssignmentIdentifier(element));
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: @inverno we can return true

File: javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java (hunk 6 above)
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/ArrayReverseCheck.java]: @inverno can remove

File: javascript-checks/src/test/resources/checks/ArrayReverse.js [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+
+  class NotArray {
+    reverse() {}
+  }
+
+  const notArray = new NotArray();
+
+  const notArrayReversed = notArray.reverse(); // OK
+
+  return a.map(b => b.reverse()); // FN, we don't know that b is an array
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/ArrayReverse.js]: @inverno remove this FN

File: javascript-checks/src/test/resources/checks/ArrayReverse.js [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+  class WithGetter {
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/ArrayReverse.js]: @inverno delete

**Pull request merged**